### PR TITLE
Don't wrap around when deriving an extended key at a too large depth

### LINKED ISSUE #25751: pubkey.cpp:368:18: runtime error: implicit conversion from type 'int' of value 256 (32-bit, signed) to type 'unsigned char' changed the value to 0 (8-bit, unsigned)
Steps to reproduce:

* Build with `integer` sanitizer
* Start: `UBSAN_OPTIONS="suppressions=$(pwd)/test/sanitizer_suppressions/ubsan:print_stacktrace=1:halt_on_error=1:report_error_type=1" ./src/qt/bitcoin-qt -chain=main`
* Enter `getdescriptorinfo "sh(multi(1,xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/5/5/5/2/8/5/8/8/8/2/2/5/8/5/2/5/85/58/5/2/5/5/5/5/2/58/2/5/8/0000/00/8/5/2/001/1/246/8/501/46/8/5/8/5/000/8/52/5/8/5/2/8/2/5/8/5/246/8/5/2/8/2/5/8/5/2/5/8/8/2/5/5/5/58/88/2/6/8/5/2/8/2/5/8/2/588/5/2/5/8/5/2/8/5/1/5/4/5/5/5/2/6/8/5/2/8/2/5/8/5/2/5/58/58/2/5/8/58/588/2/52/5/5/5/58/8/2/6/8/5/2/8/2/5/8/2/588/5/2/5/8/5/2/8/5/15/58/58/2/6/8/5/2/8/2/5/8/5/2/5/8/58/2/5/58/58/588/2/5/2/8/5/8/5/4/5/58/588/2/6/8/5/2/8/2/5/8/2/8/5/2/8/2/5/8/5/2/5/58/58/2/5/5/5/58/588/2/6/8/5/2/8/2/5/8/2/8/588/2/6/8/5/2/8/2/5/5/8/5/2/0/8/5/2/00/4/5/5/588/2/6/8/5/2/8/25/5/8/2/5/58/58/8/4/1,[00000000]xprv9uPDJpEQgRQfDcW7BkF7eTya6RPxXeJCqCJGHuCJ4GiRVLzkTXBAJMu2qaMWPrS7AANYqdq6vcBcBUdJCVVFceUvJFjaPdGZ2y9WACViL4L/58)0)"` into the console

### PR DESCRIPTION
**Pull request by darosior**
We would previously  silently wrap the derived child's depth back to `0`. Instead, explicitly fail when trying to derive an impossible depth, and handle the error in callers.

An extended fuzzing corpus of `descriptor_parse` triggered this behaviour, which was reported by MarcoFalke.

Fixes #25751.

### CODE REVIEW

**Comment by luke-jr** [comment]
Is the wrap actually harmful?

In any case, current code will never do this, right?

**Comment by darosior** [comment]
It is harmful in the sense that this behaviour is not specified by BIP32. We could accept a descriptor or an xpub that another software would refuse (or worse, treat differently).

> In any case, current code will never do this, right?

Well, it *was* triggered by current code.

**Comment by achow101** [comment]
ACK 84f86dcc1d9ff6f3161c9567068216d7b27d3b39

File: src/key.cpp [hunk 1]
```diff
@@ -333,6 +333,7 @@ bool CKey::Derive(CKey& keyChild, ChainCode &ccChild, unsigned int nChild, const
 }
 
 bool CExtKey::Derive(CExtKey &out, unsigned int _nChild) const {
```
Comment (MarcoFalke) [inline src/key.cpp:335]: Maybe mark as `[[nodiscard]]` in the header?

**Review by maflcko — COMMENTED** [review]
left a nit

File: src/key.cpp (hunk 1 above)
Comment (darosior) [reply src/key.cpp:335]: Done, and fixed two missing occurrences (in the descriptor tests) this revealed. Thanks!

**Comment by achow101** [comment]
re-ACK fb9faffae3a26b8aed8b671864ba679747163019

**Comment by instagibbs** [comment]
Hadn't occurred to me this implicit limit due to the encoding of xpub/xprv.

utACK  https://github.com/bitcoin/bitcoin/pull/25642/commits/fb9faffae3a26b8aed8b671864ba679747163019

**Pull request merged**